### PR TITLE
Use cpuinfo to get cpu instruction set extension information

### PR DESCRIPTION
**Pull request by startrekdude**
# Description

Fixes get_cpu_flags returning incorrect results for processors without AVX2 by replacing custom code with code based on [cpuinfo](https://github.com/pytorch/cpuinfo).

[cpuinfo](https://github.com/pytorch/cpuinfo) is licensed under the BSD license, which [has been confirmed](https://github.com/AOMediaCodec/SVT-AV1/pull/1596#issuecomment-732726005) to be compatible with SVT-AV1.

# Issue

Addresses #1588 

# Author(s)

Sam Haskins
(cpuinfo by cpuinfo contributors)

# Performance impact

- [ ] quality
- [ ] memory
- [x] speed
- [ ] 8 bit
- [ ] 10 bit
- [ ] N/A

# Test set
- [ ] obj-1-fast can be found [here](https://media.xiph.org/video/derf/objective-1-fast.tar.gz)
- [ ] other
- [x] N/A

# Merge method
- [x] Allow the maintainer to squash and merge when PR is ready to create a 1-commit to the master branch. The maintainer will be able to fix typos / combine commit messages to create a more readable 1-commit message or use whatever is stated in the 'Description' section
- [ ] I will clean up my commits and the maintainer shall use 'rebase and merge' to the master branch

### CODE REVIEW

**Comment by ePirat** [comment]
Hi, thanks for your work on this.
Though the amount of files and code you are adding compared to what was there before seems very overkill.

It seems this adds 300k LoC for something that was done in about 200 before. Even if the test before was not complete it seems weird that you would need that much code, and at a quick glance a big amount of the features of the cpuinfo code seems not relevant for SVT-AV1?

**Comment by hassount** [comment]
> Hi, thanks for your work on this.
> Though the amount of files and code you are adding compared to what was there before seems very overkill.
> 
> It seems this adds 300k LoC for something that was done in about 200 before. Even if the test before was not complete it seems weird that you would need that much code, and at a quick glance a big amount of the features of the cpuinfo code seems not relevant for SVT-AV1?

I agree, adding 300k lines does not seem to be the right way to go in this case as it is almost as much as the whole project has, @startrekdude would you be able to restrict it to the file / functions needed to get it working without the need of merging all of the cpuinfo in ?

**Comment by startrekdude** [comment]
Thank you both for your feedback.

Of the 300k lines, >270k were tests, and while it's certainly nice that cpuinfo has a comprehensive test suite, they don't need to be in tree—I've removed them, along with other ancillary code. The real line count of cpuinfo, as given by `cloc src --exclude-dir=arm`, is 6172 (+ 344 for clog).

**Comment by hassount** [comment]
@spawlows @ttrigui can you guys take a quick look at this ?

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -86,6 +86,7 @@ set(lib_list
     $<TARGET_OBJECTS:ENCODER_ASM_AVX2>
     $<TARGET_OBJECTS:ENCODER_ASM_AVX512>
     $<TARGET_OBJECTS:ENCODER_GLOBALS>
+    cpuinfo_public
```
Comment (hassount) [inline test/CMakeLists.txt:89]: do we still need this ?

**Comment by hassount** [comment]
> Thank you both for your feedback.
> 
> Of the 300k lines, >270k were tests, and while it's certainly nice that cpuinfo has a comprehensive test suite, they don't need to be in tree—I've removed them, along with other ancillary code. The real line count of cpuinfo, as given by `cloc src --exclude-dir=arm`, is 6172 (+ 344 for clog).

thank you @startrekdude , do you think we can reduce this further, if we don't need the arm directory that you're excluding for example, can we remove that ? it just seems that we're still far off from the 200 lines of code that we had before and that we're adding ~22k lines to fix a bug, with all due respect to the great work that you did to integrate the code and address all warnings. I am just hoping if we can find a way to take whatever is needed to address the bug in #1588

File: test/CMakeLists.txt (hunk 1 above)
Comment (startrekdude) [reply test/CMakeLists.txt:89]: Yes, just tried building the unit tests without it and it needs to be there for linking to complete.

**Comment by startrekdude** [comment]
> thank you @startrekdude , do you think we can reduce this further, if we don't need the arm directory that you're excluding for example, can we remove that ? it just seems that we're still far off from the 200 lines of code that we had before and that we're adding ~22k lines to fix a bug, with all due respect to the great work that you did to integrate the code and address all warnings. I am just hoping if we can find a way to take whatever is needed to address the bug in #1588

@hassount I've gone ahead and removed all code not used for ISA detection.

```
>cloc third_party/cpuinfo/src --ignored con
      14 text files.
      14 unique files.
third_party/cpuinfo/src: zero sized file
third_party/cpuinfo/src/cpuinfo: zero sized file
third_party/cpuinfo/src/x86: zero sized file
third_party/cpuinfo/src/x86/linux: zero sized file
third_party/cpuinfo/src/x86/mach: zero sized file
third_party/cpuinfo/src/x86/windows: zero sized file
Wrote con
       6 files ignored.

github.com/AlDanial/cloc v 1.84  T=0.50 s (28.0 files/s, 2814.0 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C                                8            149            429            521
C/C++ Header                     6             47              7            254
-------------------------------------------------------------------------------
SUM:                            14            196            436            775
-------------------------------------------------------------------------------
```

**Comment by hassount** [comment]
> > thank you @startrekdude , do you think we can reduce this further, if we don't need the arm directory that you're excluding for example, can we remove that ? it just seems that we're still far off from the 200 lines of code that we had before and that we're adding ~22k lines to fix a bug, with all due respect to the great work that you did to integrate the code and address all warnings. I am just hoping if we can find a way to take whatever is needed to address the bug in #1588
> 
> @hassount I've gone ahead and removed all code not used for ISA detection.
> 
> ```
> >cloc third_party/cpuinfo/src --ignored con
>       14 text files.
>       14 unique files.
> third_party/cpuinfo/src: zero sized file
> third_party/cpuinfo/src/cpuinfo: zero sized file
> third_party/cpuinfo/src/x86: zero sized file
> third_party/cpuinfo/src/x86/linux: zero sized file
> third_party/cpuinfo/src/x86/mach: zero sized file
> third_party/cpuinfo/src/x86/windows: zero sized file
> Wrote con
>        6 files ignored.
> 
> github.com/AlDanial/cloc v 1.84  T=0.50 s (28.0 files/s, 2814.0 lines/s)
> -------------------------------------------------------------------------------
> Language                     files          blank        comment           code
> -------------------------------------------------------------------------------
> C                                8            149            429            521
> C/C++ Header                     6             47              7            254
> -------------------------------------------------------------------------------
> SUM:                            14            196            436            775
> -------------------------------------------------------------------------------
> ```

Thank you very much @startrekdude we will proceed with the testing of the PR. in the meantime can you please confirm that you have signed the appropriate CLA as shown in the [contributor agreement](https://github.com/AOMediaCodec/SVT-AV1#contributor-agreement) section ?

**Comment by startrekdude** [comment]
> Thank you very much @startrekdude we will proceed with the testing of the PR. in the meantime can you please confirm that you have signed the appropriate CLA as shown in the [contributor agreement](https://github.com/AOMediaCodec/SVT-AV1#contributor-agreement) section ?

Yes, I've signed the Individual CLA under my name, Sam Haskins.

**Comment by hassount** [comment]
@startrekdude can you please look into why ./build.sh does not work anymore ?

**Comment by startrekdude** [comment]
> @startrekdude can you please look into why ./build.sh does not work anymore ?

@hassount It works fine for me on both my server and a fresh Ubuntu 20.04 VM. What issues are you experiencing? Maybe it's related to 9ca8333 or fe229f9?

**Comment by hassount** [comment]
> > @startrekdude can you please look into why ./build.sh does not work anymore ?
> 
> @hassount It works fine for me on both my server and a fresh Ubuntu 20.04 VM. What issues are you experiencing? Maybe it's related to [9ca8333](https://github.com/AOMediaCodec/SVT-AV1/commit/9ca833328c76bafff20736054a45210b2d8ea0f5) or [fe229f9](https://github.com/AOMediaCodec/SVT-AV1/commit/fe229f95ba166d6ff2619ecf30580f40b66e9086)?

I am using an ubuntu 18.04, running ./build.sh on your master branch, gives me the following output:

```
CMake Error at third_party/cpuinfo/CMakeLists.txt:143 (TARGET_LINK_LIBRARIES):
  Object library target "cpuinfo" may not link to anything.


-- Configuring incomplete, errors occurred!
See also "/home/user/Hassene/prs/startrekdude/Build/linux/Release/CMakeFiles/CMa                                                                                      keOutput.log".
See also "/home/user/Hassene/prs/startrekdude/Build/linux/Release/CMakeFiles/CMa                                                                                      keError.log".

```

**Comment by startrekdude** [comment]
> I am using an ubuntu 18.04, running ./build.sh on your master branch, gives me the following output:
@hassount Sorry about that, should be fixed now.

**Comment by EwoutH** [comment]
It looks like this PR is breaking non-x86 builds, see #1614. Travis is down at the moment (see https://github.com/AOMediaCodec/SVT-AV1/pull/1579#issuecomment-735688358) so that's why it didn't catch it.

**Pull request merged**